modules/pam_userdb/pam_userdb.c

```diff
@@ -213,30 +213,39 @@ user_lookup (pam_handle_t *pamh, const char *database, const char *cryptmode,
 	  } else if (ctrl & PAM_ICASE_ARG) {
 	    compare = -2;
 	  } else {
+		char *setting = malloc(data.dsize+1);
+		if (setting != NULL) {
+			strncpy(setting, data.dptr, data.dsize);
+			setting[data.dsize] = '\0';
+			if (ctrl & PAM_DEBUG_ARG) {
+				pam_syslog(pamh, LOG_INFO, "crypt setting: %s", setting);
+			}
 #ifdef HAVE_CRYPT_R
 	    struct crypt_data *cdata = NULL;
 	    cdata = malloc(sizeof(*cdata));
 	    if (cdata != NULL) {
 		cdata->initialized = 0;
-		cryptpw = crypt_r(pass, data.dptr, cdata);
-	    }
+		cryptpw = crypt_r(pass, setting, cdata);
+		free(cdata);
+		}
 #else
-	    cryptpw = crypt (pass, data.dptr);
+	    cryptpw = crypt (pass, setting);
 #endif
+		free(setting);
+	    }
 	    if (cryptpw && strlen(cryptpw) == (size_t)data.dsize) {
 	      compare = memcmp(data.dptr, cryptpw, data.dsize);
 	    } else {
 	      compare = -2;
 	      if (ctrl & PAM_DEBUG_ARG) {
-		if (cryptpw)
+		if (cryptpw) {
 		  pam_syslog(pamh, LOG_INFO, "lengths of computed and stored hashes differ");
-		else
+		  pam_syslog(pamh, LOG_INFO, "computed hash: %s", cryptpw);
+        } else {
 		  pam_syslog(pamh, LOG_INFO, "crypt() returned NULL");
+        }
 	      }
 	    }
-#ifdef HAVE_CRYPT_R
-	    free(cdata);
```

Comment:
You cannot move the free, it needs to stay here.

Reply:
Could you tell me why? It is not obvious to me and I'd like to learn more not to make similar mistakes in future.

Reply:
Just read the crypt_r manpage and what it tells about the data argument and the returned value from the crypt_r. The returned value is pointing into the cdata.